modules/apps/asset/asset-taglib/src/main/resources/META-INF/resources/asset_categories_selector/AssetVocabularyCategoriesSelector.es.js

```diff
@@ -12,300 +12,276 @@
-				.join(),
-			singleSelect: this.singleSelect,
-			vocabularyIds: this.vocabularyIds.concat()
-		});
+	useEffect(() => {
+		ref.current = value;
+	}, [value]);
 
-		const itemSelectorDialog = new ItemSelectorDialog({
-			eventName: this.eventName,
-			title: Liferay.Language.get('select-categories'),
-			url: uri
-		});
+	return ref.current;
+}
 
-		itemSelectorDialog.open();
-		itemSelectorDialog.on('selectedItemChange', event => {
-			const selectedItems = event.selectedItem;
-
-			if (selectedItems) {
-				this.selectedItems = Object.keys(selectedItems).map(itemKey => {
-					return {
-						label: selectedItems[itemKey].value,
-						value: selectedItems[itemKey].categoryId
-					};
-				});
-			}
-		});
-	}
-
-	_handleInputFocus(event) {
-		this.emit('inputFocus', event);
-	}
-
-	/**
-	 * Converts the list of selected categories into a comma-separated serialized
-	 * version to be used as a fallback for old services and implementations.
-	 *
-	 * @private
-	 * @return {string} The serialized, comma-separated version of the selected items.
-	 */
-	_getCategoryIds() {
-		return this.selectedItems
-			.map(selectedItem => selectedItem.value)
-			.join();
-	}
-
-	/**
-	 * Shows the category error.
-	 *
-	 * @private
-	 */
-	_handleErrorAddinglabelItem(event) {
-		if (event.data.itemDoesNotExists) {
-			this._typedCategory = event.target.inputValue;
-			this._unexistingCategoryError = true;
+function AssetVocabulariesCategoriesSelector({
+	eventName,
+	id,
+	isValid = true,
+	groupIds = [],
+	inputName,
+	label,
+	onSelectedItemsChange = () => {},
+	portletURL,
+	required,
+	selectedItems = [],
+	singleSelect,
+	sourceItemsVocabularyIds = [],
+	useFallbackInput
+}) {
+	const [inputValue, setInputValue] = useState('');
+	const [invalidItems, setInvalidItems] = useState([]);
+
+	const {refetch, resource} = useResource({
+		fetchOptions: {
+			body: new URLSearchParams({
+				cmd: JSON.stringify({
+					'/assetcategory/search': {
+						'-obc': null,
+						end: 20,
+						groupIds,
+						name: `%${inputValue.toLowerCase()}%`,
+						start: 0,
+						vocabularyIds: sourceItemsVocabularyIds
+					}
+				}),
+				p_auth: Liferay.authToken
+			}),
+			credentials: 'include',
+			method: 'POST',
+			'x-csrf-token': Liferay.authToken
+		},
+		link: `${window.location.origin}${themeDisplay.getPathContext()}
+				/api/jsonws/invoke`
+	});
+
+	const previousInputValue = usePrevious(inputValue);
+
+	useEffect(() => {
+		if (inputValue && inputValue !== previousInputValue) {
+			refetch();
 		}
-	}
+	}, [inputValue, previousInputValue, refetch]);
+
+	const getUnique = (arr, compKey) => {
+		return arr
+			.map(e => e[compKey])
+			.map((e, i, final) => final.indexOf(e) === i && i)
+			.filter(e => arr[e])
+			.map(e => arr[e]);
+	};
+
+	const handleItemsChange = items => {
+		const addedItems = getUnique(
+			items.filter(
+				item =>
+					!selectedItems.find(
+						selectedItem => selectedItem.value === item.value
+					)
+			),
+			'label'
+		);
```

Comment:
There is quite a bit of iteration going on here. `getUnique` itself will pass over the array 4 times (3 `map` and 1 `filter` call); additionally, inside one of those `map` calls, it does a nested iteration with `indexOf`, which makes it quadratic. Then here two we have a set of nested quadratic loops (a `find` that visits the whole array for each item visited with `filter`).

Looking further down the file I see more quadratic calls (`map` + `find`, `filter` + `find`). All of this is likely indicative of a bad choice of data structure: sets, maps, or objects would all probably be better choices.